Tech review: Add Chrome 121 new speculation rules features (#32360)
mdn/content: files/en-us/web/http/headers/no-vary-search/index.md

```

## Directives

- `key-order`
  - : A boolean. If included in the header value, it indicates that differences in the order of parameters between otherwise identical URLs will not cause them to be cached as separate entries. Differences in the parameters present _will_ cause them to be cached separately.
- `params`
  - : Either a boolean or a list of strings:
    - If included in the header value as a boolean, it indicates that differences in parameters between otherwise identical URLs will not cause them to be cached as separate entries.
    - If included in the header value as a list, it indicates that the presence of the specific parameters listed will not cause otherwise identical URLs to be cached as separate entries. The presence of other parameters _will_ cause them to be cached separately.
- `except`
  - : A list of strings. If included in the header value, it indicates that the presence of the specific parameters listed _will_ cause otherwise identical URLs to be cached as separate entries. The presence of other parameters _won't_ cause them to be cached separately. A boolean `params` directive has to be included along with `except` for it to take effect.

## Examples

> ### Stopping parameter order from causing additional cache entries

Author: OK, thanks for the explanation here @jeremyroman. I have made multiple updates to the explanation on this page, which I hope solve the problem. Let me know what you think; I may have misunderstood.

If you have for example a search page that stores its search criteria in URL parameters, and you can't guarantee that the parameters will be added to the URL in the same order each time, you can stop additional cache entries arising from URLs that are identical except for the order of the parameters using `key-order`:

> ```html

Author: Ack, I've got this wrong everywhere. Updated all of them!

No-Vary-Search: key-order
```

When added to the associated request responses, this would cause the following URLs to only produce a single cache entry:

```text
search.example.com?a=1&b=2&c=3

> search.example.com?b=2&a=1&c=3

Author: Updated

```

> The prescence of different URL parameters, however, will cause multiple cache entries. For example:

Author: Thanks, fixed!

```text
search.example.com?a=1&b=2&c=3

> search.example.com?b=2&a=1&c=3&d=4

Author: Updated

---
title: No-Vary-Search
slug: Web/HTTP/Headers/No-Vary-Search
page-type: http-header
status:
  - experimental
browser-compat: http.headers.No-Vary-Search
spec-urls: https://wicg.github.io/nav-speculation/no-vary-search.html
---

{{HTTPSidebar}}{{SeeCompatTable}}

The **`No-Vary-Search`** response header specifies a set of rules that define how a URL's query parameters will affect cache matching. These rules dictate whether the same URL with different URL parameters should be saved as separate browser cache entries.

> > **Note:** [Speculation rules](/en-US/docs/Web/API/Speculation_Rules_API) can include an `expects_no_vary_search` field, which indicates to the browser what the expected `No-Vary-Search` value will be (if any) for documents that it is receiving prefetch/preload requests for via the speculation rules. The browser can use this to determine ahead of time whether it is more useful to wait for an existing prefetch/preload to finish, or start a new fetch request when the speculation rule is matched.

Author: Darnit - I remember mixing up these two when I did the original speculation rule API docs ;-)

I've changed all appropriate instances to "prefetch/prerender". I think I'd prefer to say that to "speculative load", in case anyone lands on this section of the page without the context of what that is.

```

## Directives

- `key-order`
  - : A boolean. If included in the header value, it indicates that differences in the order of parameters between otherwise identical URLs will not cause them to be cached as separate entries. Differences in the parameters present _will_ cause them to be cached separately.
- `params`
  - : Either a boolean or a list of strings:
    - If included in the header value as a boolean, it indicates that differences in parameters between otherwise identical URLs will not cause them to be cached as separate entries.
    - If included in the header value as a list, it indicates that the presence of the specific parameters listed will not cause otherwise identical URLs to be cached as separate entries. The presence of other parameters _will_ cause them to be cached separately.
- `except`
  - : A list of strings. If included in the header value, it indicates that the presence of the specific parameters listed _will_ cause otherwise identical URLs to be cached as separate entries. The presence of other parameters _won't_ cause them to be cached separately. A boolean `params` directive has to be included along with `except` for it to take effect.

> ## Examples

Author: OK, I've added an example up the top of the examples section to cover this nice simple case. 

One other question that sprang to mind — does using `params` (and `except`) imply `key-order` for the specified params, or not? If you are specifying certain params that will not trigger additional cache entries, does them appearing in a different order still cause additional cache entries? Or would this only be true for other params appearing that are not in the `params`/`except` lists?

I reckon this is something we should clarify. I'm also wondering if it would make sense for me to include `key-order` in my `params`/`except` examples.

> ## Examples

Author: Thanks for the reply. I've included a note in the examples where `param` is first introduced:

> **Note:** If a parameter is excluded from the cache key using `params`, if it is included in the URL it will be ignored for the purposes of cache matching, regardless of where it appears in the parameter list.

Although now that you explain it, it does sound a bit obvious. Happy to remove it if you don't think it is needed.

I've decided not to include `key-order` in all the `params` examples - we probably don't want readers getting the impression that they should always be used together.

```html
No-Vary-Search: params=("id")
```

### Multiple parameters

Say you also had URL parameters that sorted the list of users on the page in ascending or descending alphabetical order, and specified the language to display the UI strings in, for example `/users?id=345&order=asc&lang=fr`.

You could get the browser to ignore all of these when considering cache matching like so:

```html
No-Vary-Search: params=("id" "order" "lang")
```

If you wanted the browser to ignore all of them _and_ any others that might be present, you could use the boolean form of `params`:

```html
No-Vary-Search: params
```

### Specifying params that _do_ trigger separate caching

Say the app behaved differently, with `/users` pointing to the main user directory landing page and `/users?id=345` pointing to a completely separate detail page for a specific user. In this case you would want the browser to ignore all the parameters mentioned above for cache matching purposes, _except_ for `id`, which would create a new cache entry each time. This can be achieved like so:

```html
No-Vary-Search: params, except=("id")
```

## Specifications

{{Specifications}}

## Browser compatibility

> {{Compat}}

Author: Correct: https://github.com/mdn/browser-compat-data/pull/22262/files#diff-b7bb4cf4cd34439bd1d5994e6bd4c767bf750c389d01b3e7daca4a02881eaef9

> {{Compat}}

Author: It's the kind of thing that we generally only call out in the compat data, unless it is genuinely by design and/or it wouldn't be likely to ever be implemented in a wider context. But from what you are saying, this doesn't sound to be the case.

## Examples

### Allowing responses from URLs with differently ordered params to match the same cache entry

If you have for example a search page that stores its search criteria in URL parameters, and you can't guarantee that the parameters will be added to the URL in the same order each time, you can allow responses from URLs that are identical except for the order of the parameters to match the same cache entry using `key-order`:

```http
No-Vary-Search: key-order
```

> When added to the associated request responses, the following would match the same cache entry:

Reviewer: Definitely better. I'm not sure how much more precise it's possible to be without ending up sounding convoluted. Maybe...

```suggestion
When this header is added to the associated responses, the following URLs would be treated as equivalent when searching the cache:
```

Author: Yup, this is really good. Updated.

```text
https://search.example.com?a=1&b=2&c=3
https://search.example.com?b=2&a=1&c=3
```

> The presence of different URL parameters, however, will cause the response not to match the same cache entry. For example:

Reviewer: nit: similar to the above, maybe...
```suggestion
The presence of different URL parameters, however, will cause these URLs to be cached separately. For example:
```

it would be possible to be slightly more precise yet, but trying to preserve the brevity of your text as much as possible

Author: Updated!

```text
https://search.example.com?a=1&b=2&c=3
https://search.example.com?b=2&a=1&c=3&d=4
```

The below examples illustrate how to control which parameters are ignored in the context of cache matching.

### Allowing responses from URLs with a different param to match the same cache entry

Consider a case where a user directory landing page, `/users`, has already been cached. An `id` parameter might be used to bring up information on a specific user, for example `/users?id=345`. Whether this URL should be considered identical for cache matching purposes depends on the behavior of the application:

- If this parameter has the effect of loading a completely new page containing the information for the specified user, then the response from this URL should be cached separately.

> - If this parameter has the effect of highlighting the specified user on the same page, and perhaps revealing a pullout panel displaying their data, then the response from this URL should match the same cache entry. This could result in performance improvements around the loading of the user pages.

Reviewer: Technically it's the request whose URL we're matching against cache entries, not the response, but either way I'd be inclined to emphasize the outcome of using the existing stored response.

```suggestion
- If this parameter has the effect of highlighting the specified user on the same page, and perhaps revealing a pullout panel displaying their data, then it would be better for the browser to use the cached response for `/users`. This could result in performance improvements around the loading of the user pages.
```

Author: Updated

If your application behaves like the second example described above, you could cause both `/users` and `/users?id=345` to be treated as identical for caching purposes via a `No-Vary-Search` header like so:

```http
No-Vary-Search: params=("id")
```

> **Note:** If a parameter is excluded from the cache key using `params`, if it is included in the URL it will be ignored for the purposes of cache matching, regardless of where it appears in the parameter list.

### Allowing responses from URLs with multiple different params to match the same cache entry

Say you also had URL parameters that sorted the list of users on the page in ascending or descending alphabetical order, and specified the language to display the UI strings in, for example `/users?id=345&order=asc&lang=fr`.

You could get the browser to ignore all of these when considering cache matching like so:

```http
No-Vary-Search: params=("id" "order" "lang")
```

If you wanted the browser to ignore all of them _and_ any others that might be present when cache matching, you could use the boolean form of `params`:

```http
No-Vary-Search: params
```

### Specifying params that _do_ cause cache matching misses

> Say the app behaved differently, with `/users` pointing to the main user directory landing page and `/users?id=345` pointing to a completely separate detail page for a specific user. In this case you would want the browser to ignore all the parameters mentioned above for cache matching purposes, _except_ for `id`, the presence of which would cause the browser to not match the `/users` cache entry and create a new one. This can be achieved like so:

Reviewer: nit: the fact that we issue the request to the server is probably more important than whether we create a new cache entry afterward (which is going to depend on the cache control headers that response has anyway, and will only impact subsequent page loads)

```suggestion
Say the app behaved differently, with `/users` pointing to the main user directory landing page and `/users?id=345` pointing to a completely separate detail page for a specific user. In this case you would want the browser to ignore all the parameters mentioned above for cache matching purposes, _except_ for `id`, the presence of which would cause the browser to not match the `/users` cache entry and request `/users?id=345` from the server. This can be achieved like so:
```

Author: Yup, looks good - updated.
